### PR TITLE
Framework and tests

### PR DESCRIPTION
**Pull request by juliehockett**

### CODE REVIEW

File: framework/framework.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+// RegisterEventFunction registers fn as an event function. The function must have two arguments, a
+// context.Context and a struct type depending on the event, and return an error. If fn has the
+// wrong signature, RegisterEventFunction panics.
+func RegisterEventFunction(path string, fn interface{}) {
+	registerEventFunction(path, fn, http.DefaultServeMux)
+}
+
+// Start serves an HTTP server with registered functions.
+func Start(port string) error {
+	if os.Getenv("K_SERVICE") == "" {
```
Comment (tbpg) [inline framework/framework.go]: Add a comment saying why this is done? May not be clear what `K_SERVICE` is.

File: framework/framework.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+// RegisterEventFunction registers fn as an event function. The function must have two arguments, a
+// context.Context and a struct type depending on the event, and return an error. If fn has the
+// wrong signature, RegisterEventFunction panics.
+func RegisterEventFunction(path string, fn interface{}) {
+	registerEventFunction(path, fn, http.DefaultServeMux)
+}
+
+// Start serves an HTTP server with registered functions.
+func Start(port string) error {
```
Comment (tbpg) [inline framework/framework.go]: Switch to `addr` (and don't do ":"+` below)?

File: README.md [hunk 3]
```diff
@@ -1 +1,188 @@
...
+
+	"cloud.google.com/go/functions/framework"
+	"example.com/hello"
+)
+
+func main() {
+	if err := framework.RegisterHTTPFunction("/", hello.HelloWorld); err != nil {
+		log.Fatalf("framework.RegisterHTTPFunction: %v\n", err)
+	}
+	port := ":8080"
```
[the remark this answers is not in the document]
Comment (juliehockett) [reply README.md]: GKE checks the PORT env var and defaults to 8080, so updated to use similar logic here.

File: README.md [hunk 4]
```diff
@@ -1 +1,188 @@
...
+
+The Functions Framework is designed to be compatible with Knative environments.
+Just build and deploy your container to a Knative environment. Note that your app needs to listen
+`PORT` environment variable per [Knative runtime contract](https://github.com/knative/serving/blob/master/docs/runtime-contract.md#inbound-network-connectivity).
+
+# Configure the Functions Framework
+
+If you're deploying to Google Cloud Functions, you don't need to worry about writing a
+`package main`. But if you want to run your function locally (e.g., for local development),
+you may want to configure the port, function to be executed and function signature type
```
[the remark this answers is not in the document]
Comment (juliehockett) [reply README.md]: This is largely taken from an initial draft that I didn't write, and I'm sad that I missed this, I always stand by the Oxford comma :(

File: framework/framework.go (hunk 2 above)
Comment (juliehockett) [reply framework/framework.go]: Existing framework uses PORT without the semicolon: https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/master/src/index.ts#L67

File: README.md [hunk 5]
```diff
@@ -1 +1,189 @@
...
+
+func eventFunction(ctx context.Context, e myEventType){
+	// function logic
+}
+```
+
+> Note that the first parameter to a function that handles events has to be `context.Context`
+and the type of second parameter needs to be a type of an unmarshallable event.
+
+# Handle Cloud Events
```
Comment (grant) [inline README.md]: # Enable CloudEvents

**Pull request merged**